common/src/main/java/com/wynntils/gui/render/MapRenderer.java

```diff
@@ -99,10 +116,226 @@ public static void renderMapQuad(
+
+                FontRenderer.getInstance()
+                        .renderText(
+                                poseStack,
+                                labelPoi.getLabel().getName(),
+                                textureXPosition,
+                                textureZPosition,
+                                0,
+                                CommonColors.WHITE,
+                                HorizontalAlignment.Left,
+                                VerticalAlignment.Top,
+                                FontRenderer.TextShadow.NORMAL);
+            }
+        }
+
         // disable rotation if necessary
         if (followPlayerRotation) {
             poseStack.popPose();
         }
+
+        renderServicePois(
+                poseStack,
+                mouseCoordinates,
+                mapCenterX,
+                mapCenterZ,
+                centerX,
+                centerZ,
+                scale,
+                poiScale,
+                followPlayerRotation,
+                servicePois);
+    }
+
+    private static boolean isPoiVisible(
+            float mapCenterX,
+            float mapCenterZ,
+            float centerX,
+            float centerZ,
+            float scale,
+            float mapTopX,
+            float mapLeftZ,
+            float mapBottomX,
+            float mapRightZ,
+            Poi poi) {
+        float textureXPosition = poi.getRenderX(mapCenterX, centerX, 1f / scale);
+        float textureZPosition = poi.getRenderZ(mapCenterZ, centerZ, 1f / scale);
+
+        return textureXPosition >= mapTopX
+                && textureXPosition <= mapBottomX
+                && textureZPosition >= mapLeftZ
+                && textureZPosition <= mapRightZ;
+    }
+
+    private static void renderServicePois(
+            PoseStack poseStack,
+            Pair<Integer, Integer> mouseCoordinates,
+            float mapCenterX,
+            float mapCenterZ,
+            float centerX,
+            float centerZ,
+            float scale,
+            float poiScale,
+            boolean followPlayerRotation,
+            List<ServicePoi> servicePois) {
+
+        if (mouseCoordinates != null) {
+            hovered = null;
+        }
+
+        for (ServicePoi servicePoi : servicePois) {
+            renderServicePoi(
+                    poseStack,
+                    mouseCoordinates,
+                    mapCenterX,
+                    mapCenterZ,
+                    centerX,
+                    centerZ,
+                    followPlayerRotation,
+                    scale,
+                    poiScale,
+                    servicePoi);
+        }
+
+        if (McUtils.mc().screen instanceof MainMapScreen mainMapScreen) {
+            mainMapScreen.setHovered(hovered);
+        }
+
+        if (hovered != null && mouseCoordinates != null) {
+            float textureXPosition = hovered.getRenderX(mapCenterX, centerX, 1f / scale);
```

Comment:
How can this happen? If mouseCoordinates is != null we set hovered to null first thing in the method. I can't see how this ever can be true?

Reply:
Ah, I see, we set hovered in renderServicePoi. Not completely obvious.

Reply:
This check might not be needed, and is there just to prevent weirdness.